Whitepaper/narrative voice (#109)
cncf/tag-app-delivery: operator-wg/whitepaper/02_Foundation.md

* Our domain-specific knowledge that describes the state of the application in a declarative way.
* A controller that run continuously:
  * Reads and is aware of the state.
  * Runs actions against the application in an automated way.
  * Reports the state of the application  in a declarative way.

This design pattern will be applied to Kubernetes and its Operators in the next sections.

### Operator Characteristics
**(Current) Issue: https://github.com/cncf/sig-app-delivery/issues/37**

The core purpose of any operator is to extend its framework's underlying API with new domain knowledge. As an example, an orchestration platform as Kubernetes natively understands things like containers and layer 4 load balancers via the Pod and Service objects. An operator adds new capabilities for more complex systems and applications. For instance, a Prometheus operator introduces new object types, "Prometheus", which extend Kubernetes with high-level support for deploying and running Prometheus servers.

The capabilities provided by an operator can be sorted into three overarching categories: dynamic configuration, operational automation, and domain knowledge.

#### Dynamic Configuration

> Since the dawn of time, there have been two main ways to configure software: configuration files and environment variables. In the cloud-native world there are some newer options like querying a well-known API at startup, but most existing software in the world uses one or both of these options.

Reviewer: ```suggestion
Since the early origins of modern computing, there have been two main ways to configure software; configuration files and environment variables. In the cloud-native world, there are some newer options like querying a well-known API at startup, but most existing software in the world uses one or both of these options.
```

It was a bit cheesy

Kubernetes naturally provides many tools to interact with these (ConfigMaps, Secrets, etc.) but because they are generic, they don’t understand any specifics of configuring a given application. An operator can define new custom object types (Custom Resources) to better express the configuration of a particular application in a Kubernetes context.

This allows for better validation and data structuring, which in turn reduces the likelihood of small configuration errors and improves the ability of teams to self-service without having as deep or complete a knowledge of either the underlying framework or the target application as would be traditionally required. This can include things like progressive defaults, where a few high-level settings are used to populate an opinionated or best-practices-driven configuration file or adaptive configuration such as adjusting resource usage to match available hardware or expected load based on cluster size.

#### Operational Automation

> Along with custom resources, most operators include at least one custom controller. These controllers are daemons that run inside the orchestrator like any other, but connect to the underlying API and provide automation of common or repetitive tasks. This is the same way that orchestrators (like Kubernetes) are implemented, as you may have seen kube-controller-manager or cloud-controller-manager mentioned in your journey so far. But as we saw with configuration, Operators can extend and enhance orchestrators with higher-level automation such as deploying clustered software, providing automated backups and restores, or dynamic scaling based on load.

Reviewer: ```suggestion
Along with custom resources, most operators include at least one custom controller. These controllers are daemons that run inside the orchestrator like any other, but connect to the underlying API and provide automation of common or repetitive tasks. This is the same way that orchestrators like Kubernetes are implemented (kube-controller-manager or cloud-controller-manager mentioned often mentioned as familiar landmarks). But as we saw with configuration, Operators can extend and enhance orchestrators with higher-level automation such as deploying clustered software, providing automated backups and restores, or dynamic scaling based on load.
```
